packages/vite/src/node/plugins/dynamicImportVars.ts

```diff
@@ -0,0 +1,214 @@
+import { posix } from 'path'
+import MagicString from 'magic-string'
+import { init, parse as parseImports } from 'es-module-lexer'
+import type { ImportSpecifier } from 'es-module-lexer'
+import type { Plugin } from '../plugin'
+import type { ResolvedConfig } from '../config'
+import { normalizePath, parseRequest } from '../utils'
+import { parse as parseJS } from 'acorn'
+import { createFilter } from '@rollup/pluginutils'
+import { dynamicImportToGlob } from '@rollup/plugin-dynamic-import-vars'
+
+export const dynamicImportHelperId = '/@vite/dynamic-import-helper'
+
+interface DynamicImportRequest {
+  as?: 'raw'
+}
+
+interface DynamicImportPattern {
+  globParams: DynamicImportRequest | null
+  userPattern: string
+  rawPattern: string
+}
+
+const dynamicImportHelper = (glob: Record<string, any>, path: string) => {
+  const v = glob[path]
+  if (v) {
+    return typeof v === 'function' ? v() : Promise.resolve(v)
+  }
+  return new Promise((_, reject) => {
+    ;(typeof queueMicrotask === 'function' ? queueMicrotask : setTimeout)(
+      reject.bind(null, new Error('Unknown variable dynamic import: ' + path))
+    )
+  })
+}
+
+function parseDynamicImportPattern(
+  strings: string
+): DynamicImportPattern | null {
+  const filename = strings.slice(1, -1)
+  const rawQuery = parseRequest(filename)
```

Comment:
Parsing the request here could be fragile if `filename` is something like `../alias/${base ?? 'foo'}.js?raw`. Maybe we should have a hardcoded check for endsWith `?raw` and `?url`?

Reply:
I think we can match the query from end to start?
```js
const res = filename.split('?')
const pattern = res.slice(0, res.length - 2).join('')
const query = res[-1]
```

Reply:
how about regexp https://regex101.com/r/I9TKTV/1 😊

Reply:
Oo that seems like it could work. I think that'll do! Nice find.